spread/openstack.go

```diff
@@ -0,0 +1,1030 @@
+}
+
+func (p *openstackProvider) GarbageCollect() error {
+	if err := p.checkKey(); err != nil {
+		return err
+	}
+
+	instances, err := p.listServers()
+	if err != nil {
+		return err
+	}
+
+	volumes, err := p.listVolumes()
+	if err != nil {
+		return err
+	}
+	snapshots, err := p.listSnapshots()
+	if err != nil {
+		return err
+	}
+
+	now := time.Now()
+	haltTimeout := p.backend.HaltTimeout.Duration
+
+	// Iterate over all the running instances
+	for _, s := range instances {
+		serverTimeout := haltTimeout
+		if value, ok := s.d.Labels["halt-timeout"]; ok {
+			d, err := time.ParseDuration(strings.TrimSpace(value))
+			if err != nil {
+				printf("WARNING: Ignoring bad openstack instances %s halt-timeout label: %q", s, value)
+			} else {
+				serverTimeout = d
+			}
+		}
+
+		if serverTimeout == 0 {
+			continue
+		}
+
+		printf("Checking openstack instance %s...", s)
+
+		runningTime := now.Sub(s.d.Created)
+		if runningTime > serverTimeout {
+			printf("Server %s exceeds halt-timeout. Shutting it down...", s)
+			err := p.removeMachine(context.Background(), s)
+			if err != nil {
+				printf("WARNING: Cannot garbage collect server %s: %v", s, err)
+			}
+		}
+	}
+
+	// Iterate over all the volumes
+	for _, v := range volumes {
+		printf("Checking openstack volume %s...", v.Id)
+
+		snapshotAttached := ""
+		for _, s := range snapshots {
+			if s.VolumeID == v.Id {
+				snapshotAttached = s.Id
+				break
+			}
+		}
+
+		if snapshotAttached == "" {
+			printf("Volume ready to be deleted %s. Shutting it down...", v.Id)
+			err := p.volumeClient.DeleteVolume(v.Id)
+			if err != nil {
+				printf("WARNING: Cannot garbage collect volume %s: %v", v.Id, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (p *openstackProvider) saveServices() error {
+	endpoints := p.osClient.EndpointsForRegion(p.region)
+	p.services = &openstackServices{}
+
+	for k, v := range endpoints {
+		if strings.HasPrefix(k, "volume") || strings.HasPrefix(k, "compute") {
+			ver := "v2"
+			if strings.HasSuffix(k, "v3") {
+				ver = "v3"
+			}
+			endpointUrl, err := url.Parse(v)
+			if err != nil {
+				return &FatalError{fmt.Errorf("error parsing endpoint: %v", &openstackError{err})}
+			}
+
+			service := OpenstackService{
+				Name:     k,
+				version:  ver,
+				endpoint: endpointUrl,
+			}
+			if strings.HasPrefix(k, "compute") {
+				p.services.compute = service
+			} else {
+				p.services.volume = service
+			}
+		}
+	}
+	if p.services.compute.Name == "" {
+		return &FatalError{fmt.Errorf("compute services endpoint not found")}
+	}
+	if p.services.volume.Name == "" {
+		return &FatalError{fmt.Errorf("volume services endpoint not found")}
+	}
+	return nil
+}
+
+func (p *openstackProvider) checkKey() error {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	if p.keyChecked {
+		return p.keyErr
+	}
+
+	var err error
+
+	if err == nil && p.computeClient == nil {
+
+		// Load environment variables used to authenticate
+		if p.backend.Key != "" {
+			godotenv.Load(p.backend.Key)
+		}
```

Comment:
Here we're getting a file defining environment variables and reading it into the environment (to be read from the environment later). could be nicer to extract the variables from the file and consume them directly (so they won't be accessible from the environment), or source the file before running spread and maybe dropping this field altogether?